Adding a tool to demultiplex FASTQs.

nh13 opened the pull request.

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+  }
+
+  /** Increments the metrics for the given sample ordinal, which matched with the given number of mismatches.  The
+    * sample ordinal less than zero indicates the "unmatched" sample. */
```
less than zero not true anymore -> remove

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+    case n => throw new IllegalArgumentException(s"Found $n read structures with template bases but expected 1 or 2.")
+  }
+
+  /** The concatenation of all read structures read structure. */
```
remove extra `read structure`

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+  /** The number of reads that are expected to be given to the [[demultiplex()]]. */
+  def expectedNumberOfReads: Int = readStructures.length
+
+  /** True if we the read structure implies paired end reads will be produced, false otherwise. */
```
change to: `True if the read structure implies paired end reads will be produced, false otherwise`

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+  require(headers.length == samples.length + 1, s"Not enough SAMFileHeaders; found ${headers.length} but should have ${samples.length+1}.")
+  require(headers.forall(_.getReadGroups.size == 1), "All headers must have one read group. Ordinals are: " + headers.zipWithIndex.filter(_._1.getReadGroups.size != 1).map(_._2).mkString(", "))
+
+  /** The number of reads that are expected to be given to the [[demultiplex()]]. */
```
add `method` to the end of the sentence.

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+      |  - Sample Barcode:     Index
+      |The expected sample barcode should be placed in the 'Index' column.  If the sample barcode is present across
+      |multiple reads (ex. dual-index, or inline in both reads of a pair), then the expected barcode bases from each read
+      |should be concatenated and placed in teh 'Index' column.  The concatenation should be in the same order as the
```
teh typo

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+  }
+
+    /** Gets the sample ordinal given the bases from all cycles and the number of mismatches between the bases and matched
+      * sample barcode.  If no match is found, the ordinal for the oridinal for the 'unmatched" sample is given. */
```
indentation.

nh13 replied:
the ordinal for the oridinal for the 'unmatched" -> the "unmatched" ordinal

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+    require(reads.nonEmpty, "No reads given for demultiplexing.")
+    require(reads.length == expectedNumberOfReads, s"Expected '$expectedNumberOfReads' number of reads but found '${reads.length}'.")
+
+    // Get the sample ordinal
```
_sample index_

nh13 commented on src/test/scala/com/fulcrumgenomics/fastq/DemuxFastqsTest.scala:
```diff
@@ -0,0 +1,336 @@
+
+import scala.collection.mutable.ListBuffer
+
+class DemuxFastqsTest extends UnitSpec with OptionValues with ErrorLogLevel {
```
@tfenne I really like both the `OptionValues` and `ErrorLogLevel` traits.  The former allows us to replace the `get` method with the `value` method for `Option`s and scalatest report a meaningful error when the `Option` is `None`.  The latter allows us to turn off logging to the console when running end-to-end `tool.execute()` tests.

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+                                 val readStructures: Seq[ReadStructure],
+                                 val umiTag: String = "RX",
+                                 val qualityFormat: FastqQualityFormat = FastqQualityFormat.Standard,
+                                 val minQ: Int = 0,
```
I feel like enforcement of the min/max quality doesn't belong here.  This is really validation that the fastqs aren't borked, and surely that should reside with the thing reading the fastqs?  Perhaps we should add options to FastqSource to translate qualities and enforce the quality range, so that it's re-useable elsewhere?

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+  val pairedEnd: Boolean = readStructures.count(_.template.nonEmpty) match {
+    case 1 => false
+    case 2 => true
+    case n => throw new IllegalArgumentException(s"Found $n read structures with template bases but expected 1 or 2.")
```
Move this check up into DemuxFastqs?

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+  }
+
+  /** The concatenation of all read structures read structure. */
+  private val fullReadStructure: ReadStructure = new ReadStructure(segs=readStructures.flatten, resetOffsets=true)
```
I see where this is going, but I think there's a problem.  I routinely see data generated on MiSeqs where the fastqs were generated with the option to trim adapters.  The result is that the template reads are not fixed length, and thus you can't just concatenate the read structures and templates like this.

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+  private val fullReadStructure: ReadStructure = new ReadStructure(segs=readStructures.flatten, resetOffsets=true)
+
+  /** The quality format converter. */
+  private val solexaQualityConverter: SolexaQualityConverter = SolexaQualityConverter.getSingleton
```
The `SolexaQualityConverter` is pretty icky for a bunch of reasons.  I wonder if it would make sense to port it to scala (it's only about a dozen lines of actual code) and also offer `String => String` versions of the methods?

nh13 replied:
Another day.

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+
+  /** Get the sample barcode bases across all bases. */
+  private def sampleBarcodeBases(allBases: String): String = {
+    ReadStructure.structureRead(bases=allBases, segments=sampleBarcodeSegments).map(_.bases).mkString
```
I think I see why there are both, but it would be nice to note why you have this and also the one below.  Could even just be a single method that takes in the delimiter so you can use it both ways?

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+          .getOrElse(unreachable(s"Sample with id '${sample.sampleId}' did not have a sample barcode"))
+          .barcodeBytes
+        val numMismatches = countMismatches(observedBarcode, expectedBarcode)
+        (sample.sampleOrdinal-1, numMismatches)
```
Why `sampleOrdinal - 1`?

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+      }
+    }
+    else {
+      (this.samples.size, Int.MaxValue, Int.MaxValue)
```
It would be nice to extract `this.samples.size` into a class level val called `UnmatchedSampleOrdinal` and use that everywhere to make it clear what's going on.

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+    // Create the SAMRecords.
+    val header            = headers(sampleIndex)
+    val sampleId          = header.getReadGroups.get(0).getId
+    val molecularBarcode  = this.molecularBarcode(allBases)
```
prefixing method names with `this` is generally frowned upon

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+        val subReads: Seq[SubRead] = ReadStructure.structureReadWithQualities(bases=read.bases, qualities=read.quals, segments=templateSegments)
+        val bases                  = subReads.map(_.bases).mkString
+        val quals                  = subReads.flatMap(_.quals).mkString
+        require(bases.length == quals.length, s"Bases and qualities have differing lengths forread: ${read.header}")
```
forread -> for read

tfenne commented:
Thinking about my point 4 a bit more, I think you could literally create a single case class:
`case class SampleInfo(sample: Sample: writer: SAMFileWriter, header: SAMFileHeader, rg: SAMReadGroupRecord)`, make a `Seq` of those, and pass the `Seq` around to things that need to see all the samples, and a `SampleInfo` anywhere you need just one, then all the lookups could be avoided without too much surgery.

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+  }
+
+  /** Gets the quality format of the FASTQs. */
+  private def determineQualityFormat(fastqs: Seq[PathToFastq], qualityFormat: Option[FastqQualityFormat] = None, logger: Option[Logger] = None): FastqQualityFormat = {
```
It's actually the expected quality format.  If it's `None`, it guesses the best format, and if it is `Some(format)` then it verifies the quality format.  This is the same behavior as Picard's `FastqToSam`, which I am trying to keep a little bit of.

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+  group=ClpGroups.Fastq
+)
+class DemuxFastqs
+(@arg(flag="f", doc="One or more input fastq files each corresponding to a sub-read (ex. index read, read one, read two, fragment).") val fastqs:  Seq[PathToFastq],
```
If you have multiple FASTQs, then yes, you will need to run this tool once per group of FASTQs, or at least combine them prior to running this tool.  As a by-product, the former allows for simple parallelization in a pipeline!

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+  /** Counts the nucleotide mismatches between two strings of the same length.  Ignores no calls in expectedBases.
+    * Observed base qualities less than the minimum base quality are counted as mismatches if not a no call.
+    */
+  private[fastq] def countMismatches(observedBases: Array[Byte], expectedBases: Array[Byte]): Int = {
```
The optimization in `matchSampleBarcode` is not to even check against _any_ sample barcodes if there are too many no calls in the observed bases.

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,614 @@
+  private[fastq] def countMismatches(observedBases: Array[Byte], expectedBases: Array[Byte]): Int = {
+    var idx = 0
+    var count = 0
+    while (idx < observedBases.length) {
```
I like _while_ every once in a while.

nh13 commented:
@tfenne I think I addressed all your concerns, and made a note where I did not, which were few and far between.  I think `Sample` and `SampleSheet` could be further improved, but I running out of steam to do that here.

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,573 @@
+
+    // Create the SAMRecords.
+    val header            = sampleInfo.header
+    val sampleId          = sampleInfo.rg.getId
```
This should probably be `rgId`.

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,573 @@
+        if (pairedEnd) {
+          record.setReadPairedFlag(true)
+          record.setMateUnmappedFlag(true)
+          if (readNumber == 1) {
```
This could be:
```scala
readNumber match {
  case 1 => record.setFirstOfPairFlag(true)
  case 2 => record.setSecondOfPairFlag(true) 
  case n => throw new IllegalStateException(s"Invalid read number '$n' for read: ${read.header}")
}
```

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,573 @@
+        }
+        record.setAttribute("BC", sampleBarcode)
+        record.setAttribute(ReservedTagConstants.READ_GROUP_ID, sampleId)
+        if (molecularBarcode.nonEmpty) record.setAttribute("RX", molecularBarcode)
```
I believe this should be `this.umiTag` not `"RX"`.

tfenne commented on src/main/scala/com/fulcrumgenomics/illumina/Sample.scala:
```diff
@@ -42,22 +46,30 @@ package com.fulcrumgenomics.util.miseq
+  lazy val sampleBarcodeBytes: Array[Byte]  = sampleBarcodes.flatten.filter(_.nonEmpty).flatMap(_.getBytes).toArray
+  lazy val sampleBarcodeString: String      = sampleBarcodes.flatten.filter(_.nonEmpty).mkString(SampleBarcodeDelimiter)
+  override def equals(o: Any): Boolean = o.isInstanceOf[Sample] && this.sampleBarcodeString == o.asInstanceOf[Sample].sampleBarcodeString
+  override def hashCode: Int           = this.sampleBarcodeString.hashCode
```
I'm not sure about this.  This essentially means that if you have a sample sheet that doesn't have sample barcodes defined, then the hashCode will be `"".hashCode` which seems wrong.  Is the goal here to just have something faster than the standard case class `hashCode`?  Maybe you could `+ sampleOrdinal`?

nh13 replied:
Ditto about the copy-paste error.

tfenne commented on src/main/scala/com/fulcrumgenomics/illumina/Sample.scala:
```diff
@@ -42,22 +46,30 @@ package com.fulcrumgenomics.util.miseq
+
+  lazy val sampleBarcodeBytes: Array[Byte]  = sampleBarcodes.flatten.filter(_.nonEmpty).flatMap(_.getBytes).toArray
+  lazy val sampleBarcodeString: String      = sampleBarcodes.flatten.filter(_.nonEmpty).mkString(SampleBarcodeDelimiter)
+  override def equals(o: Any): Boolean = o.isInstanceOf[Sample] && this.sampleBarcodeString == o.asInstanceOf[Sample].sampleBarcodeString
```
Why not leave the default case class implementation of equals?  This seems over-simplified.

nh13 replied:
Copy and past error from the previous `SampleBarcode` class.  Thanks for catching this.

tfenne commented on src/main/scala/com/fulcrumgenomics/illumina/Sample.scala:
```diff
@@ -42,22 +46,30 @@ package com.fulcrumgenomics.util.miseq
+  lazy val sampleBarcodeString: String      = sampleBarcodes.flatten.filter(_.nonEmpty).mkString(SampleBarcodeDelimiter)
+  override def equals(o: Any): Boolean = o.isInstanceOf[Sample] && this.sampleBarcodeString == o.asInstanceOf[Sample].sampleBarcodeString
+  override def hashCode: Int           = this.sampleBarcodeString.hashCode
+  override def toString: String        = this.sampleBarcodeString
```
Same comment here.  This seems very specific to your use in DemuxFastqs.  Perhaps the toString could be at least `sampleName/sampleBarcodeString`?

tfenne commented on src/test/scala/com/fulcrumgenomics/illumina/SampleTest.scala:
```diff
@@ -54,6 +54,22 @@ class SampleTest extends UnitSpec with OptionValues{
+    an[Exception] should be thrownBy new Sample(0, "ID", "NAME", "LIBRARY", extendedAttributes=Seq(("foo", "1"), ("bar", "2  ")).toMap)
+  }
+
+  "Sample.hashCode" should "be the hash code of the concatenated barcode sequences" in {
```
I think it would be better to write these tests to test the desired behaviour (e.g. hashCodes don't result in collisions for different samples) rather than the specific implementation.

tfenne commented on src/test/scala/com/fulcrumgenomics/fastq/DemuxFastqsTest.scala:
```diff
@@ -0,0 +1,524 @@
+      |,,,,,,,,,
+      |[Data],,,,,,,,,
+      |Sample_ID,Sample_Name,Sample_Plate,Sample_Well,index,index2,Sample_Barcode,Sample_Project,Description
+      |20000101-EXPID-1,Sample_Name_1,,,,,AAAAAAAAGATTACAGA,Sample_Project_1,Description_1
```
Since you use these elsewhere it might be nice to extract the barcode sequences into four vals and use `${barcode1}` syntax to insert them here.  Then you can refer to them in tests without hard-coding them in multiple places.

tfenne commented on src/test/scala/com/fulcrumgenomics/fastq/DemuxFastqsTest.scala:
```diff
@@ -0,0 +1,524 @@
+    r1.getBaseQualityString shouldBe "I"*100
+    r1.getReadPairedFlag shouldBe true
+    r1.getFirstOfPairFlag shouldBe true
+    r1.getReadUnmappedFlag shouldBe true
```
to be on the safe side you could also assert that `getSecondOfPairFlag` is `false`.

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -0,0 +1,573 @@
+  /** Converts the segments across all reads to a string, with each read delimited by the given delimiter. */
+  private def toBarcode(allBases: Seq[String], delimiter: String, allSegments: Seq[Seq[ReadSegment]]): String = {
+    allBases.zip(allSegments).map { case (bases, segments) =>
+      ReadStructure.structureRead(bases=bases, segments=segments).map(_.bases).filter(_.nonEmpty).mkString
```
The barcodes are delimited by read, not segment.
